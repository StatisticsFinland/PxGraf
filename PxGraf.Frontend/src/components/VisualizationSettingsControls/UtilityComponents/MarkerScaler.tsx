import { Box, Stack, Typography, Slider } from '@mui/material';
import React from 'react';
import { useTranslation } from 'react-i18next';
import { IVisualizationSettingsProps } from '../VisualizationSettingsControl';
import styled from 'styled-components';

const FormControlWrapper = styled(Stack)`
    gap: 40px;
    align-items: center;
`;

const StyledSlider = styled(Slider)`
    min-width: 150px;
    margin-right: 16px;
`;

export const MarkerScaler: React.FC<IVisualizationSettingsProps> = ({ visualizationSettings, settingsChangedHandler }) => {
    const { t } = useTranslation();
    // We dont want to trigger the api call every time the slider moves, only when the change is committed.
    const [sliderValue, setSliderValue] = React.useState(visualizationSettings.markerSize);

    const marks = [
        {
            value: 1,
            label: '1%',
        },
        {
            value: 100,
            label: '100%',
        },
        {
            value: 200,
            label: '200%',
        }
    ];

    return (
        <FormControlWrapper direction="row">
            <Box>
                <Typography id="input-slider" gutterBottom>
                    {t("chartSettings.markerScale")}
                </Typography>
                <StyledSlider
                    min={1}
                    max={200}
                    marks={marks}
                    value={sliderValue}
                    onChange={(_event, value) => setSliderValue(typeof (value) == "number" ? value : 100)}
                    onChangeCommitted={(_event, value) => settingsChangedHandler({ ...visualizationSettings, markerSize: typeof (value) == "number" ? value : 100 })}
                    aria-label={t("chartSettings.markerScaleDefaultLabel")}
                    valueLabelDisplay="auto" />
            </Box>
        </FormControlWrapper>

    );
}

export default MarkerScaler;