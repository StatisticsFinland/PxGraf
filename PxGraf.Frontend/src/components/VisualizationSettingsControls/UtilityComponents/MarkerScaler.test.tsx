import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import MarkerScaler from "./MarkerScaler";

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => null),
            },
        };
    },
}));

const mockVisualizationRules = {
    allowManualPivot: null,
    sortingOptions: null,
    multiselectVariableAllowed: null
};

const mockVisualizationSettings = {
    defaultSelectableVariableCodes: null,
    pivotRequested: null,
    cutYAxis: false,
    multiselectableVariableCode: null,
    rowVariableCodes: null,
    columnVariableCodes: null,
    sorting: null,
    matchXLabelsToEnd: null,
    markerSize: 100,
}

describe('Rendering test', () => {
    it('renders correctly', () => {
        const mockSettingsChangedHandler = jest.fn();
        const { asFragment } = render(
            <MarkerScaler
                visualizationRules={mockVisualizationRules}
                visualizationSettings={mockVisualizationSettings}
                settingsChangedHandler={mockSettingsChangedHandler}
            />);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Assertion tests', () => {
    it('Marker size value should be 123 when the corresponding parameter value is 123', () => {
        const mockSettingsChangedHandler = jest.fn();
        const toggledMockSettings = { ...mockVisualizationSettings, markerSize: 123 }
        render(
            <MarkerScaler
                visualizationRules={mockVisualizationRules}
                visualizationSettings={toggledMockSettings}
                settingsChangedHandler={mockSettingsChangedHandler}
            />);
        expect(screen.getByDisplayValue('123')).toHaveValue('123');
    });

    it('MarkerScaler onChange and onChangeCommitted should work properly', () => {
        const mockSettingsChangedHandler = jest.fn();
        render(
            <MarkerScaler
                visualizationRules={mockVisualizationRules}
                visualizationSettings={mockVisualizationSettings}
                settingsChangedHandler={mockSettingsChangedHandler}
            />
        );

        const slider = screen.getByRole('slider');
        fireEvent.change(slider, { target: { value: 50 } });
        expect(slider).toHaveValue('50');

        fireEvent.mouseUp(slider);
        expect(mockSettingsChangedHandler).toHaveBeenCalledWith({ ...mockVisualizationSettings, markerSize: 50 });
    });
});