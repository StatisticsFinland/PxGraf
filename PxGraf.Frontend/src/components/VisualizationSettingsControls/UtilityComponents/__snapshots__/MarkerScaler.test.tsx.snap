// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiStack-root sc-braxZu MhTIp css-m69qwo-MuiStack-root"
  >
    <div
      class="MuiBox-root css-0"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-d3wcwz-MuiTypography-root"
        id="input-slider"
      >
        chartSettings.markerScale
      </p>
      <span
        class="MuiSlider-root MuiSlider-marked MuiSlider-colorPrimary MuiSlider-sizeMedium sc-gJhJTp evpvpa css-ltlhnc-MuiSlider-root"
      >
        <span
          class="MuiSlider-rail css-7o8aqz-MuiSlider-rail"
        />
        <span
          class="MuiSlider-track css-ttgsjq-MuiSlider-track"
          style="left: 0%; width: 49.74874371859296%;"
        />
        <span
          class="MuiSlider-mark MuiSlider-markActive css-z5sdnj-MuiSlider-mark"
          data-index="0"
          style="left: 0%;"
        />
        <span
          aria-hidden="true"
          class="MuiSlider-markLabel MuiSlider-markLabel MuiSlider-markLabelActive css-1ucwjgd-MuiSlider-markLabel"
          data-index="0"
          style="left: 0%;"
        >
          1%
        </span>
        <span
          class="MuiSlider-mark MuiSlider-markActive css-z5sdnj-MuiSlider-mark"
          data-index="1"
          style="left: 49.74874371859296%;"
        />
        <span
          aria-hidden="true"
          class="MuiSlider-markLabel MuiSlider-markLabel MuiSlider-markLabelActive css-1ucwjgd-MuiSlider-markLabel"
          data-index="1"
          style="left: 49.74874371859296%;"
        >
          100%
        </span>
        <span
          class="MuiSlider-mark css-hzp7sc-MuiSlider-mark"
          data-index="2"
          style="left: 100%;"
        />
        <span
          aria-hidden="true"
          class="MuiSlider-markLabel MuiSlider-markLabel css-1bvr8oc-MuiSlider-markLabel"
          data-index="2"
          style="left: 100%;"
        >
          200%
        </span>
        <span
          class="MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeMedium MuiSlider-thumbColorPrimary css-cp2j25-MuiSlider-thumb"
          data-index="0"
          style="left: 49.74874371859296%;"
        >
          <input
            aria-label="chartSettings.markerScaleDefaultLabel"
            aria-orientation="horizontal"
            aria-valuemax="200"
            aria-valuemin="1"
            aria-valuenow="100"
            data-index="0"
            max="200"
            min="1"
            step="1"
            style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
            type="range"
            value="100"
          />
          <span
            aria-hidden="true"
            class="MuiSlider-valueLabel css-14gyywz-MuiSlider-valueLabel"
          >
            <span
              class="MuiSlider-valueLabelCircle"
            >
              <span
                class="MuiSlider-valueLabelLabel"
              >
                100
              </span>
            </span>
          </span>
        </span>
      </span>
    </div>
  </div>
</DocumentFragment>
`;
